### PR TITLE
BUG-19214 int categoricals are formatted as ints

### LINKED ISSUE #19214: categorical series with null converts ints to float
```python
In [4]: import pandas as pd, numpy as np

In [5]: sr = pd.Series([1,2, np.nan], dtype="O").astype("category")
   ...: sr
   ...: 
Out[5]: 
0    1.0
1    2.0
2    NaN
dtype: category
Categories (2, int64): [1, 2]

In [6]: sr[0]
Out[6]: 1.0
```
Shouldn't it keep them as ints? Interestingly, `pd.Series(["a", 1,2, np.nan], dtype="O").astype("category")` doesn't do the conversion.

### PR DESCRIPTION
**Pull request by JustinZhengBC**
- [X] closes #19214
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

This PR alters the `ExtensionArrayFormatter` such that if it receives a categorical series with only integers and `NaN`, it will print the integers as integers instead of converting to floats.

```
>>> # previous behaviour
>>> pd.Series([1,2, np.nan], dtype="object").astype("category")
0    1.0
1    2.0
2    NaN
dtype: category
Categories (2, int64): [1, 2]
>>> # new behaviour
>>> pd.Series([1,2, np.nan], dtype="object").astype("category")
0      1
1      2
2    NaN
dtype: category
Categories (2, int64): [1, 2]
```

### CODE REVIEW

File: pandas/io/formats/format.py [hunk 1]
```diff
@@ -1176,6 +1176,11 @@ def _format_strings(self):
                                   na_rep=self.na_rep, digits=self.digits,
                                   space=self.space, justify=self.justify,
                                   leading_space=self.leading_space)
+
```
Comment (jreback) [inline pandas/io/formats/format.py]: we have a CategoricalFormatter already, mod should be made there. Should be caught when formatting, not as a post-action.

File: pandas/tests/arrays/categorical/test_repr.py [hunk 2]
```diff
@@ -240,6 +240,10 @@ def test_categorical_repr_datetime_ordered(self):
 
         assert repr(c) == exp
 
+    def test_categorical_repr_int_with_nan(self):
+        s = Series([1, 2, np.nan], dtype="object").astype("category")
+        assert ".0" not in repr(s)
```
Comment (jreback) [inline pandas/tests/arrays/categorical/test_repr.py]: assert the exact repr (similar to other tests)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by JustinZhengBC** [comment]
@jreback I looked into it and the `CategoricalFormatter` is used for printing categoricals like this:
```
[1.0, 2.0, NaN]
Categories (2, int64): [1, 2]
```
instead of like a series. The PR now modifies the `CategoricalFormatter` in addition to the `ExtensionArrayFormatter`.

File: pandas/io/formats/format.py [hunk 3]
```diff
@@ -134,8 +134,11 @@ def _get_footer(self):
         return compat.text_type(footer)
 
     def _get_formatted_values(self):
-        return format_array(self.categorical.get_values(), None,
-                            float_format=None, na_rep=self.na_rep)
+        results = format_array(self.categorical.get_values(), None,
+                               float_format=None, na_rep=self.na_rep)
+        if is_integer_dtype(self.categorical.dtype.categories):
```
Comment (jreback) [inline pandas/io/formats/format.py]: ok again the issue is that this is not how we need to do this.

**Review by jreback — CHANGES_REQUESTED** [review]
So the fix is in ``.get_values()`` do this

```
n [28]: c.categories.astype(object).take(c.codes, fill_value=np.nan)
Out[28]: Index([1, 2, nan], dtype='object')
```

if -1 in c.codes and  is_integer_dtype(c.categories)

**Comment by jreback** [comment]
@TomAugspurger any objections here?

File: doc/source/whatsnew/v0.24.0.rst [hunk 4]
```diff
@@ -1321,6 +1321,7 @@ Categorical
 - Bug in many methods of the ``.str``-accessor, which always failed on calling the ``CategoricalIndex.str`` constructor (:issue:`23555`, :issue:`23556`)
 - Bug in :meth:`Series.where` losing the categorical dtype for categorical data (:issue:`24077`)
 - Bug in :meth:`Categorical.apply` where ``NaN`` values could be handled unpredictably. They now remain unchanged (:issue:`24241`)
+- Bug in :meth:`Categorical.get_values` where integers would be coerced to floats if ``NaN`` values were present (:issue:`19214`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.rst]: can you make this a user facing message; mention both the formatting fix & the concat changes

File: pandas/core/arrays/categorical.py [hunk 5]
```diff
@@ -1520,6 +1520,9 @@ def get_values(self):
         # if we are a datetime and period index, return Index to keep metadata
         if is_datetimelike(self.categories):
             return self.categories.take(self._codes, fill_value=np.nan)
+        elif -1 in self._codes and is_integer_dtype(self.categories):
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: reverse this check (check for integer first)

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]
the warnings are not needed

File: doc/source/whatsnew/v0.24.0.rst [hunk 6]
```diff
@@ -1323,6 +1323,7 @@ Categorical
 - Bug in many methods of the ``.str``-accessor, which always failed on calling the ``CategoricalIndex.str`` constructor (:issue:`23555`, :issue:`23556`)
 - Bug in :meth:`Series.where` losing the categorical dtype for categorical data (:issue:`24077`)
 - Bug in :meth:`Categorical.apply` where ``NaN`` values could be handled unpredictably. They now remain unchanged (:issue:`24241`)
+- Bug in :meth:`Categorical.get_values` where integers would be formatted as floats if ``NaN`` values were present (:issue:`19214`)
```
[the remark this answers is not in the document]
Comment (JustinZhengBC) [reply doc/source/whatsnew/v0.24.0.rst]: @jreback I'm a bit confused as to what a user-facing note is. Is it a specific section somewhere? Or does it mean to phrase the note differently, like "bug in the Categorical repr"?

[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.rst]: sorry. what I mean is a user would want to know that the string repr of Series (or categorical dtype) and Categoricals for integer categories will now not be coerced to float. Its maybe worth making this a sub-section (e.g. show previous / new behavior). you can just in-line it here.

File: doc/source/whatsnew/v0.24.0.rst [hunk 7]
```diff
@@ -1653,6 +1683,7 @@ Reshaping
 - :meth:`DataFrame.nlargest` and :meth:`DataFrame.nsmallest` now returns the correct n values when keep != 'all' also when tied on the first columns (:issue:`22752`)
 - Constructing a DataFrame with an index argument that wasn't already an instance of :class:`~pandas.core.Index` was broken (:issue:`22227`).
 - Bug in :class:`DataFrame` prevented list subclasses to be used to construction (:issue:`21226`)
+- Calling :func:`pandas.concat` on a ``Categorical`` of ints with NA values now causes them to be processed as objects (formerly coerced to floats) (:issue:`19214`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.rst]: This only applies when concating a categorical with a different dtype, right? If I concat two integer cats with the same dtype, it’s still categoricalmright?

File: pandas/tests/arrays/categorical/test_repr.py [hunk 8]
```diff
@@ -240,6 +240,15 @@ def test_categorical_repr_datetime_ordered(self):
 
         assert repr(c) == exp
 
+    def test_categorical_repr_int_with_nan(self):
+        c = Categorical([1, 2, np.nan])
+        c_exp = """[1, 2, NaN]\nCategories (2, int64): [1, 2]"""
+        assert repr(c) == c_exp
+
+        s = Series([1, 2, np.nan], dtype="object").astype("category")
+        s_exp = """0      1\n1      2\n2    NaN\ndtype: category\nCategories (2, int64): [1, 2]"""  # noqa
```
Comment (TomAugspurger) [inline pandas/tests/arrays/categorical/test_repr.py]: You can use parenthesis to wrap this line.

**Review by TomAugspurger — COMMENTED** [review]
To me, the change in dtype for concat is the larger change. Just from the release notes it looks like the formatting change is larger.

**Comment by jreback** [comment]
> To me, the change in dtype for concat is the larger change. Just from the release notes it looks like the formatting change is larger.

I don't think this is that big of a deal, do you want a sub-section for this?

**Comment by jreback** [comment]
@TomAugspurger you think need to highlite the change more?

**Comment by TomAugspurger** [comment]
The formatting change seems fine. I'm more concerned about the breaking
change to get_values
with no deprecation warning. Is there any way to deprecate the behavior, or
would that require to
many internal changes so that the warning doesn't show up when printing the
categorical?

**Comment by jreback** [comment]
> I'm more concerned about the breaking
change to get_values
with no deprecation warning. Is there any way to deprecate the behavior, or
would that require to

This is a very small case to be honest (and our current behavior for concatting any non-same types, so this is a bug IMHO). 

@JustinZhengBC can you make a small sub-section about the changes in Categorical int-like concat, showing previous and current behavior. I think that is fine.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@TomAugspurger better?

File: doc/source/whatsnew/v0.24.0.rst [hunk 9]
```diff
@@ -1140,6 +1140,39 @@ cast from integer dtype to floating dtype (:issue:`22019`)
        ...:                    'c': [1, 1, np.nan, 1, 1]})
     In [4]: pd.crosstab(df.a, df.b, normalize='columns')
 
+.. _whatsnew_0240.api.concat_categorical:
+
+Concatenation Changes
+^^^^^^^^^^^^^^^^^^^^^
+
+Calling :func:`pandas.concat` on a ``Categorical`` of ints with NA values now
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.rst]: Specific that this only applies when concatenating a Categorical with something of a different dtype.

File: doc/source/whatsnew/v0.24.0.rst [hunk 10]
```diff
@@ -1140,6 +1140,39 @@ cast from integer dtype to floating dtype (:issue:`22019`)
...
+causes them to be processed as objects in cases where previously they would be
+coerced to floats (:issue:`19214`)
+
+*Previous Behavior*
+
+.. code-block:: ipython
+    In [2]: s = pd.Series([0, 1, np.nan])
+
+    In [3]: c = pd.Series([0, 1, np.nan], dtype="object").astype("category")
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.rst]: Why not `c = pd.Series([0, 1, np.nan], dtype="category")`?

**Review by TomAugspurger — COMMENTED** [review]
Just doc comments.

**Comment by jreback** [comment]
@TomAugspurger ?

**Review by TomAugspurger — APPROVED** [review]

**Comment by jreback** [comment]
thanks @JustinZhengBC keep em coming!

**Pull request merged**